feat: improve extensibility of forgot password feature

markywilson opened the pull request:
closes #10496

markywilson committed:
feat: Move update email feature to user library (GH-9631)

Closes #9788 

Co-authored-by: tobi-or-not-tobi <tobi-or-not-tobi@users.noreply.github.com>

markywilson committed:
Merge branch 'develop' of github.com:SAP/spartacus into feature/GH-9630

markywilson committed:
(move) all methods and form declaration to ForgotPassword Service

markywilson committed:
Merge branch 'develop' of github.com:SAP/spartacus into feature/GH-9630

tobi-or-not-tobi commented on projects/storefrontlib/src/cms-components/myaccount/forgot-password/forgot-password.component.ts:
```diff
@@ -1,48 +1,17 @@
+import { ForgotPasswordService } from './forgot-password.service';
 @Component({
   selector: 'cx-forgot-password',
   templateUrl: './forgot-password.component.html',
```
can you add on push change detection here?

tobi-or-not-tobi commented on projects/storefrontlib/src/cms-components/myaccount/forgot-password/forgot-password.component.html:
```diff
@@ -1,25 +1,18 @@
+    <cx-form-errors [control]="form.get('userEmail')"></cx-form-errors>
+  </label>
 
   <button type="submit" class="btn btn-block btn-primary">
```
nitpick: submit is default, so you could drop that actually

tobi-or-not-tobi commented on projects/storefrontlib/src/cms-components/myaccount/forgot-password/forgot-password.service.ts:
```diff
@@ -0,0 +1,51 @@
+})
+export class ForgotPasswordService {
+  constructor(
+    protected fb: FormBuilder,
```
form builder is unused and indeed not needed

tobi-or-not-tobi commented on projects/storefrontlib/src/cms-components/myaccount/forgot-password/forgot-password.service.ts:
```diff
@@ -0,0 +1,51 @@
+  });
+
+  submit() {
+    if (!this.form.valid) {
```
or `if (this.form.invalid)` :D

tobi-or-not-tobi commented on projects/storefrontlib/src/cms-components/myaccount/forgot-password/forgot-password.service.ts:
```diff
@@ -0,0 +1,51 @@
+    }
+  }
+
+  reset() {
```
I'd either make the form an internal property (protected) or do a form.reset() in the controller directly.

markywilson replied:
I did it like that because I thought the `reset` method might need to do more stuff in the future not only resetting the form (resetting the state for example).

tobi-or-not-tobi replied:
ah, that sounds fair

tobi-or-not-tobi reviewed (changes requested):
small things only, good stuff!

markywilson committed:
Merge branch 'develop' of github.com:SAP/spartacus into feature/GH-9630

markywilson committed:
Merge branch 'feature/GH-9630' of github.com:SAP/spartacus into feature/GH-10496

markywilson committed:
(change) PR comments

tobi-or-not-tobi reviewed (approved):
LGTM

(merging on hold though, as we need to plan the breaking changes)

tobi-or-not-tobi committed:
Merge branch 'feature/GH-9630-user-lib-state-services' into feature/GH-10496

tobi-or-not-tobi committed:
move refactored component to new user lib

tobi-or-not-tobi committed:
Merge branch 'develop' into feature/GH-10496

tobi-or-not-tobi committed:
Adjust logic to new user lib structure

tobi-or-not-tobi committed:
Merge branch 'feature/GH-9630-user-lib-state-services' into feature/GH-10496

tobi-or-not-tobi committed:
move email form component to user lib and revert previous refactor in original location

tobi-or-not-tobi committed:
add deprecations and revert changes in obsolete storefront comp

tobi-or-not-tobi committed:
deprecations

tobi-or-not-tobi committed:
revert tests

tobi-or-not-tobi committed:
revert test

tobi-or-not-tobi committed:
fix styling

tobi-or-not-tobi committed:
Add login page template to user lib

tobi-or-not-tobi committed:
Merge branch 'feature/GH-9630-user-lib-state-services' into feature/GH-10496

tobi-or-not-tobi committed:
fix tests

tobi-or-not-tobi committed:
Move user styles to library

dunqan commented on feature-libs/user/profile/components/forgot-password/forgot-password.service.ts:
```diff
@@ -0,0 +1,75 @@
+        );
+      });
+
+    this.redirect();
```
Shouldn't we redirect after requestForgotPaswordEmail will finish? (in subscribe?)

tobi-or-not-tobi committed:
use anchor link instead of button

dunqan commented on feature-libs/user/profile/components/update-email/update-email.module.ts:
```diff
@@ -1,22 +1,28 @@
 import { UpdateEmailComponent } from './update-email.component';
 
+/**
+ * @deprecated since 3.2, moved to @spartacus/user package.
```
Hmmm... aren't we in the user lib here?

dunqan commented on feature-libs/user/profile/components/update-email/update-email.service.spec.ts:
```diff
@@ -0,0 +1,172 @@
+  update(_password: string, _newUid: string): Observable<unknown> {
+    return of({});
+  }
+  // updateEmail(): void {}
```
left overs?

tobi-or-not-tobi replied:
yes already fixed

tobi-or-not-tobi committed:
Merge branch 'feature/GH-9630-user-lib-state-services' into feature/GH-10496

tobi-or-not-tobi committed:
review fixes

tobi-or-not-tobi committed:
refactor update password component

tobi-or-not-tobi committed:
Refactor profile feature

tobi-or-not-tobi committed:
Refactor

- forgot password
- reset password
- improve tests
- make implementation and tests consistent

tobi-or-not-tobi committed:
cleanup PR

tobi-or-not-tobi committed:
Merge branch 'feature/GH-9630-user-lib-state-services' into feature/GH-10496

tobi-or-not-tobi committed:
revert components other then forget library

tobi-or-not-tobi committed:
revert css changes for other components

tobi-or-not-tobi committed:
missing selector in user lib styles

tobi-or-not-tobi committed:
make component services configurable
